File: qr-pdf/generate_qr_pdf.py
#!/usr/bin/env python3
"""Generate a print-ready A4 PDF of QR codes for the main-library games.

One QR code per game in the collection, each pointing at that game's
learn-to-play short link (https://shiny-hoppy-meeple.pages.dev/learn-to-play/<slug>/),
with the game name and full address printed beneath it. Lay it out on A4, print,
cut along the guides, and stick the codes on the boxes.

Usage:
    python generate_qr_pdf.py                       # defaults: main library → qr-codes.pdf
    python generate_qr_pdf.py --output stickers.pdf --cols 3 --rows 4
    python generate_qr_pdf.py --scan-slugs ../shiny-hoppy-meeple/public/scan-slugs.json

The slug is derived from the game name with the same algorithm Hugo's `anchorize`
uses (GitHub heading-ID rules), so the generated links match the site's URLs. Pass
--scan-slugs to cross-check every slug against the site's generated allowlist.
"""

import argparse
import json
import sys
from pathlib import Path

import segno
from reportlab.lib.pagesizes import A4
from reportlab.lib.units import mm
from reportlab.pdfgen import canvas

DEFAULT_COLLECTION = (
    Path(__file__).parent.parent
    / "shiny-hoppy-meeple" / "data" / "bgg-cache" / "collections" / "main-library.json"
)
DEFAULT_BASE_URL = "https://shiny-hoppy-meeple.pages.dev/"
DEFAULT_PATH_PREFIX = "learn-to-play"
DEFAULT_OUTPUT = (
    Path(__file__).parent.parent / "shiny-hoppy-meeple" / "static" / "qr-codes.pdf"
)

PAGE_MARGIN = 10 * mm   # outer margin around the grid
CELL_PADDING = 4 * mm   # inner padding within each grid cell
QR_QUIET_ZONE = 2       # QR modules of white border around the code


def anchorize(name: str) -> str:
    """Slugify a game name to match Hugo's `anchorize` (GitHub heading-ID rules).

    Lowercase; keep letters, numbers and hyphens; turn each space into a hyphen;
    drop everything else. Consecutive hyphens are preserved (not collapsed), so
    "Rome & Roll" → "rome--roll", matching the slugs Hugo writes for the site.
    """
    out = []
    for ch in name.lower():
        if ch == " ":
            out.append("-")
        elif ch == "-" or ch.isalnum():
            out.append(ch)
        # all other characters are dropped
    return "".join(out)


def load_games(collection_path: Path, base_url: str, path_prefix: str) -> list[dict]:
    """Read the collection JSON and return [{name, slug, url}] for each game."""
    if not collection_path.exists():
        sys.exit(f"Error: collection file not found: {collection_path}")

    data = json.loads(collection_path.read_text())
    items = data.get("items", [])
    base = base_url.rstrip("/")
    prefix = path_prefix.strip("/")

    games = []
    for item in items:
        name = item.get("name", "")
        if not name:
            continue
        slug = anchorize(name)
        games.append({
            "name": name,
            "slug": slug,
            "url": f"{base}/{prefix}/{slug}/",
        })
    return games


def cross_check_slugs(games: list[dict], scan_slugs_path: Path) -> None:
    """Warn if any generated slug is missing from the site's scan-slugs allowlist."""
    if not scan_slugs_path.exists():
        print(f"  Warning: scan-slugs file not found, skipping check: {scan_slugs_path}",
              file=sys.stderr)
        return

    allowed = set(json.loads(scan_slugs_path.read_text()))
    missing = [g for g in games if g["slug"] not in allowed]
    if missing:
        print(f"  Warning: {len(missing)} slug(s) not found in {scan_slugs_path.name}:",
              file=sys.stderr)
        for g in missing:
            print(f"    {g['name']!r} → {g['slug']}", file=sys.stderr)
    else:
        print(f"  Slug check: all {len(games)} slugs present in {scan_slugs_path.name}")


def _fit_font_size(c: canvas.Canvas, text: str, font: str, max_width: float,
                   start: float, minimum: float = 5.0) -> float:
    """Largest font size (down to `minimum`) at which `text` fits in `max_width`."""
    size = start
    while size > minimum and c.stringWidth(text, font, size) > max_width:
        size -= 0.5
    return size


def draw_qr(c: canvas.Canvas, matrix, x: float, y: float, size: float) -> None:
    """Draw a segno QR `matrix` as filled vector squares inside a `size` box at (x, y)."""
    rows = list(matrix)
    n = len(rows)
    modules = n + 2 * QR_QUIET_ZONE          # include the quiet zone
    scale = size / modules
    origin = QR_QUIET_ZONE * scale

    c.setFillColorRGB(0, 0, 0)
    for r, row in enumerate(rows):
        for col, dark in enumerate(row):
            if dark:
                # PDF y grows upward; draw rows top-to-bottom.
                px = x + origin + col * scale
                py = y + size - origin - (r + 1) * scale
                c.rect(px, py, scale, scale, stroke=0, fill=1)


def draw_cell(c: canvas.Canvas, game: dict, x: float, y: float,
              width: float, height: float) -> None:
    """Render one game cell: cut border, QR, bold name, and the URL beneath."""
    # Cut guide.
    c.setStrokeColorRGB(0.8, 0.8, 0.8)
    c.setLineWidth(0.5)
    c.rect(x, y, width, height, stroke=1, fill=0)

    inner_w = width - 2 * CELL_PADDING
    name_size = 9
    url_size = 6
    name_h = name_size + 2
    url_h = url_size + 2
    text_block = name_h + url_h

    # QR fills the cell above the text block, constrained to a square.
    qr_area_h = height - 2 * CELL_PADDING - text_block
    qr_size = min(inner_w, qr_area_h)
    qr_x = x + (width - qr_size) / 2
    qr_y = y + CELL_PADDING + text_block + (qr_area_h - qr_size)

    qr = segno.make(game["url"], error="m")
    draw_qr(c, qr.matrix, qr_x, qr_y, qr_size)

    # Game name (bold), shrunk to fit the cell width.
    c.setFillColorRGB(0, 0, 0)
    cx = x + width / 2
    ns = _fit_font_size(c, game["name"], "Helvetica-Bold", inner_w, name_size)
    c.setFont("Helvetica-Bold", ns)
    c.drawCentredString(cx, y + CELL_PADDING + url_h, game["name"])

    # Full URL (smaller), shrunk to fit.
    us = _fit_font_size(c, game["url"], "Helvetica", inner_w, url_size, minimum=4.0)
    c.setFont("Helvetica", us)
    c.setFillColorRGB(0.3, 0.3, 0.3)
    c.drawCentredString(cx, y + CELL_PADDING, game["url"])


def build_pdf(games: list[dict], output: Path, cols: int, rows: int) -> int:
    """Lay out the games in a cols×rows grid across A4 pages. Returns page count."""
    page_w, page_h = A4
    output.parent.mkdir(parents=True, exist_ok=True)
    c = canvas.Canvas(str(output), pagesize=A4)

    grid_w = page_w - 2 * PAGE_MARGIN
    grid_h = page_h - 2 * PAGE_MARGIN
    cell_w = grid_w / cols
    cell_h = grid_h / rows
    per_page = cols * rows

    pages = 0
    for i, game in enumerate(games):
        slot = i % per_page
        if slot == 0:
            if i > 0:
                c.showPage()
            pages += 1
        row = slot // cols
        col = slot % cols
        # Top-left origin for cells; PDF y grows upward.
        x = PAGE_MARGIN + col * cell_w
        y = page_h - PAGE_MARGIN - (row + 1) * cell_h
        draw_cell(c, game, x, y, cell_w, cell_h)

    if games:
        c.showPage()
    c.save()
    return pages


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Generate a printable A4 PDF of learn-to-play QR codes for the main library."
    )
    parser.add_argument("--collection", type=Path, default=DEFAULT_COLLECTION,
                        help="Collection JSON to read games from "
                             "(default: main-library.json in the Hugo data dir)")
    parser.add_argument("--base-url", default=DEFAULT_BASE_URL,
                        help=f"Site base URL (default: {DEFAULT_BASE_URL})")
    parser.add_argument("--path-prefix", default=DEFAULT_PATH_PREFIX,
                        help=f"URL path prefix before the slug (default: {DEFAULT_PATH_PREFIX})")
    parser.add_argument("--output", type=Path, default=DEFAULT_OUTPUT,
                        help="Output PDF path "
                             "(default: shiny-hoppy-meeple/static/qr-codes.pdf, served at /qr-codes.pdf)")
    parser.add_argument("--cols", type=int, default=3, help="Columns per page (default: 3)")
    parser.add_argument("--rows", type=int, default=4, help="Rows per page (default: 4)")
    parser.add_argument("--scan-slugs", type=Path, default=None,
                        help="Optional scan-slugs.json to cross-check generated slugs against")
    args = parser.parse_args()

    games = load_games(args.collection, args.base_url, args.path_prefix)
    if not games:
        sys.exit(f"Error: no games found in {args.collection}")

    if args.scan_slugs:
        cross_check_slugs(games, args.scan_slugs)

    pages = build_pdf(games, args.output, args.cols, args.rows)
    print(f"  {len(games)} games → {pages} page(s) → {args.output}")
    print("Done.")


if __name__ == "__main__":
    main()

File: scripts/generate-qr-pdf.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const { parseArgs } = require('util');
const PDFDocument = require('pdfkit');
const QRCode = require('qrcode');

const MM = 2.8346; // 1 mm in points
const PAGE_MARGIN = 10 * MM;
const CELL_PADDING = 4 * MM;
const QR_QUIET_ZONE = 2; // modules of white border around the code

const DEFAULT_COLLECTION = path.resolve(
  __dirname, '../shiny-hoppy-meeple/data/bgg-cache/collections/main-library.json'
);
const DEFAULT_BASE_URL = 'https://shiny-hoppy-meeple.pages.dev/';
const DEFAULT_PATH_PREFIX = 'learn-to-play';
const DEFAULT_OUTPUT = path.resolve(
  __dirname, '../shiny-hoppy-meeple/static/qr-codes.pdf'
);

// Mirror of Hugo's `anchorize` (GitHub heading-ID rules), which is what produces
// the site's /games/<slug>/ URLs. Drift here prints stickers pointing at pages
// that don't exist, so the rules below were read back out of Hugo 0.163.1 and are
// pinned by test/anchorize.test.js: keep unicode letters and decimal digits,
// keep `-` and `_`, turn a space into `-`, drop everything else (punctuation,
// symbols, emoji, combining marks, non-decimal numerals like ² or Ⅳ), after
// trimming the ends.
const SLUG_KEEP = /[\p{L}\p{Nd}]/u;

function anchorize(name) {
  let out = '';
  for (const ch of name.trim().toLowerCase()) {
    if (ch === ' ') out += '-';
    else if (ch === '-' || ch === '_' || SLUG_KEEP.test(ch)) out += ch;
  }
  return out;
}

function loadGames(collectionPath, baseUrl, pathPrefix) {
  if (!fs.existsSync(collectionPath)) {
    process.stderr.write(`Error: collection file not found: ${collectionPath}\n`);
    process.exit(1);
  }
  const data = JSON.parse(fs.readFileSync(collectionPath, 'utf8'));
  const items = data.items || [];
  const base = baseUrl.replace(/\/$/, '');
  const prefix = pathPrefix.replace(/^\/|\/$/g, '');

  const games = [];
  for (const item of items) {
    const name = item.name || '';
    if (!name) continue;
    const slug = anchorize(name);
    games.push({ name, slug, url: `${base}/${prefix}/${slug}/` });
  }
  return games;
}

// Fatal by design: a slug the site doesn't serve means a printed sticker that
// 404s, and these sheets get cut up and stuck on boxes. Callers ask for this
// check explicitly with --scan-slugs, so a missing allowlist is a failure too —
// it means the check didn't happen, not that it passed. Runs before the PDF is
// written, so a bad sheet never reaches disk.
function crossCheckSlugs(games, scanSlugsPath) {
  if (!fs.existsSync(scanSlugsPath)) {
    throw new Error(`scan-slugs file not found, cannot verify slugs: ${scanSlugsPath}`);
  }
  const allowed = new Set(JSON.parse(fs.readFileSync(scanSlugsPath, 'utf8')));
  const missing = games.filter(g => !allowed.has(g.slug));
  if (missing.length) {
    for (const g of missing) {
      process.stderr.write(`    '${g.name}' → ${g.slug}\n`);
    }
    throw new Error(
      `${missing.length} slug(s) above are not in ${path.basename(scanSlugsPath)} — ` +
      'their QR codes would 404. Rebuild the site, or finalise the game names first.'
    );
  }
  process.stdout.write(
    `  Slug check: all ${games.length} slugs present in ${path.basename(scanSlugsPath)}\n`
  );
}

function fitFontSize(doc, text, maxWidth, startSize, minimum = 5.0) {
  let size = startSize;
  doc.fontSize(size);
  while (size > minimum && doc.widthOfString(text) > maxWidth) {
    size -= 0.5;
    doc.fontSize(size);
  }
  return size;
}

function drawQr(doc, qr, qrX, qrY, boxSize) {
  const { data, size } = qr.modules;
  const modules = size + 2 * QR_QUIET_ZONE;
  const scale = boxSize / modules;
  const origin = QR_QUIET_ZONE * scale;

  doc.fillColor('#000000');
  for (let r = 0; r < size; r++) {
    for (let c = 0; c < size; c++) {
      if (data[r * size + c]) {
        const px = qrX + origin + c * scale;
        const py = qrY + origin + r * scale;
        doc.rect(px, py, scale, scale).fill();
      }
    }
  }
}

function drawCell(doc, game, x, y, width, height) {
  // Cut guide
  doc.lineWidth(0.5).rect(x, y, width, height).stroke('#cccccc');

  const innerW = width - 2 * CELL_PADDING;
  const nameSize = 9;
  const urlSize = 6;
  const nameH = nameSize + 2;
  const urlH = urlSize + 2;
  const textBlock = nameH + urlH;

  // QR fills the cell above the text block, constrained to a square
  const qrAreaH = height - 2 * CELL_PADDING - textBlock;
  const qrSize = Math.min(innerW, qrAreaH);
  const qrX = x + (width - qrSize) / 2;
  const qrY = y + CELL_PADDING;

  const qr = QRCode.create(game.url, { errorCorrectionLevel: 'M' });
  drawQr(doc, qr, qrX, qrY, qrSize);

  // Game name (bold), shrunk to fit inner width
  doc.font('Helvetica-Bold').fontSize(nameSize);
  fitFontSize(doc, game.name, innerW, nameSize);
  const nameY = y + height - CELL_PADDING - urlH - nameSize;
  doc.fillColor('#000000').text(game.name, x, nameY, {
    align: 'center',
    width,
    lineBreak: false,
  });

  // Full URL (smaller), shrunk to fit; minimum 4pt
  doc.font('Helvetica').fontSize(urlSize);
  fitFontSize(doc, game.url, innerW, urlSize, 4.0);
  const urlY = y + height - CELL_PADDING - urlSize;
  doc.fillColor('#4d4d4d').text(game.url, x, urlY, {
    align: 'center',
    width,
    lineBreak: false,
  });
}

async function buildPdf(games, outputPath, cols, rows) {
  fs.mkdirSync(path.dirname(outputPath), { recursive: true });

  const doc = new PDFDocument({ size: 'A4', autoFirstPage: true, margin: 0 });
  const stream = fs.createWriteStream(outputPath);
  doc.pipe(stream);

  const pageW = doc.page.width;
  const pageH = doc.page.height;
  const gridW = pageW - 2 * PAGE_MARGIN;
  const gridH = pageH - 2 * PAGE_MARGIN;
  const cellW = gridW / cols;
  const cellH = gridH / rows;
  const perPage = cols * rows;

  let pages = 0;
  for (let i = 0; i < games.length; i++) {
    const slot = i % perPage;
    if (slot === 0) {
      if (i > 0) doc.addPage();
      pages++;
    }
    const row = Math.floor(slot / cols);
    const col = slot % cols;
    const x = PAGE_MARGIN + col * cellW;
    const y = PAGE_MARGIN + row * cellH;
    drawCell(doc, games[i], x, y, cellW, cellH);
  }

  doc.end();
  await new Promise((resolve, reject) => {
    stream.on('finish', resolve);
    stream.on('error', reject);
  });
  return pages;
}

async function main() {
  const { values } = parseArgs({
    options: {
      collection:    { type: 'string' },
      'base-url':    { type: 'string' },
      'path-prefix': { type: 'string' },
      output:        { type: 'string' },
      cols:          { type: 'string' },
      rows:          { type: 'string' },
      'scan-slugs':  { type: 'string' },
      help:          { type: 'boolean', short: 'h', default: false },
    },
  });

  if (values.help) {
    process.stdout.write([
      'Usage: node generate_qr_pdf.js [options]',
      '',
      'Options:',
      '  --collection <path>    Collection JSON (default: main-library.json in Hugo data dir)',
      '  --base-url <url>       Site base URL (default: https://shiny-hoppy-meeple.pages.dev/)',
      '  --path-prefix <str>    URL path prefix before slug (default: learn-to-play)',
      '  --output <path>        Output PDF path (default: shiny-hoppy-meeple/static/qr-codes.pdf)',
      '  --cols <n>             Columns per page (default: 3)',
      '  --rows <n>             Rows per page (default: 4)',
      '  --scan-slugs <path>    scan-slugs.json to cross-check generated slugs against',
      '  -h, --help             Show this help',
      '',
    ].join('\n'));
    process.exit(0);
  }

  const collectionPath = values.collection ?? DEFAULT_COLLECTION;
  const baseUrl = values['base-url'] ?? DEFAULT_BASE_URL;
  const pathPrefix = values['path-prefix'] ?? DEFAULT_PATH_PREFIX;
  const outputPath = values.output ?? DEFAULT_OUTPUT;
  const cols = parseInt(values.cols ?? '3', 10);
  const rows = parseInt(values.rows ?? '4', 10);

  const games = loadGames(collectionPath, baseUrl, pathPrefix);
  if (games.length === 0) {
    process.stderr.write(`Error: no games found in ${collectionPath}\n`);
    process.exit(1);
  }

  if (values['scan-slugs']) {
    crossCheckSlugs(games, values['scan-slugs']);
  }

  const pages = await buildPdf(games, outputPath, cols, rows);
  process.stdout.write(`  ${games.length} games → ${pages} page(s) → ${outputPath}\n`);
  process.stdout.write('Done.\n');
}

if (require.main === module) {
  main().catch(err => {
    process.stderr.write(`${err.message}\n`);
    process.exit(1);
  });
}

module.exports = { anchorize, crossCheckSlugs };
